#! /bin/sh /usr/share/dpatch/dpatch-run
## 04_fix_fullpath_buffer_overflow.dpatch by Jose Carlos Medeiros <debian@psabs.com.br>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix  "fullpath()" File Name Handling Buffer Overflow, CAN-2006-0855

[ -f debian/patches/00patch-opts ] && . debian/patches/00patch-opts
patch_opts="${patch_opts:--f --no-backup-if-mismatch ${2:+-d $2}}"

if [ $# -lt 1 ]; then
  echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
  exit 1
fi

case "$1" in
  -patch) patch $patch_opts -p1 < $0;;
  -unpatch) patch $patch_opts -p1 -R < $0;;
  *)
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1;;
esac

exit 0

@DPATCH@
diff -urNad zoo-2.10~/misc.c zoo-2.10/misc.c
--- zoo-2.10~/misc.c	2006-03-03 18:51:31.000000000 -0300
+++ zoo-2.10/misc.c	2006-03-03 18:54:29.000000000 -0300
@@ -136,11 +136,14 @@
 char *fullpath (direntry)
 struct direntry *direntry;
 {
-	static char result[PATHSIZE];
+	static char result[PATHSIZE+PATHSIZE+12]; // Room for enough space
 	combine (result,
 				direntry->dirlen != 0 ? direntry->dirname : "", 
 				(direntry->namlen != 0) ? direntry->lfname : direntry->fname
 			  );
+	if (strlen (result) >= PATHSIZE) {
+		prterror ('f', "Combined dirname and filename too long\n"); 
+	}
 	return (result);
 }
 
